title: "Plotting complex gene expression data"
output: rmarkdown::html_vignette
author: "Giuseppe D'Agostino"
date: "3/29/2022"
vignette: >
  %\VignetteIndexEntry{coldcuts}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

In this vignette we show how to add and plot data from a complex dataset, i.e. a dataset where structures are sampled at different timepoints or in different individuals, so that the user may want to choose additional variables before plotting. We also introduce the `seg_feature_complex_plot` function to create a rich visualization for the whole dataset.

## Getting data

For this example we will use the Allen Brain Institute Developmental Transcriptome (also known as "Brainspan"), available [here](https://www.brainspan.org/static/download.html).

We will use the data contained in *RNA-Seq Gencode v10 summarized to genes*. 

The compressed archive contains:
- normalized (RPKM) gene expression values for every time point and every individual
- row metadata (gene names, IDs and symbols)
- column metadata (patients, time points, structures)

We will combine everything together in a coldcuts `assay` and we will add it to the [ABA Human Half](https://langleylab.github.io/coldcuts/articles/segmentations.html#aba-human-half-and-full) segmentation, together with a new projection on the sagittal plane for the structures analyzed in Brainspan.

## Creating the mapping

In order to do that, we need to create a table that allows us to map structures between the segmentation and the dataset, which we need to do manually:

|name|acronym|segmentation_id|segmentation_name|segmenation_acronym|
| --- | --- | --- | --- | --- | 
|occipital neocortex|	Ocx|	NA|	NA|	NA|
primary motor-sensory cortex (samples)|	M1C-S1C|	NA|	NA|	NA|
amygdaloid complex|	AMY|	10361|	amygdaloid complex|	AMY|
medial ganglionic eminence|	MGE|	NA|	NA|	NA|
posterior (caudal) superior temporal cortex (area 22c)|	STC|	12140|	superior temporal gyrus|	STG|
upper (rostral) rhombic lip|	URL|	NA|	NA|	NA|
caudal ganglionic eminence|	CGE|	NA|	NA|	NA|
dorsal thalamus|	DTH|	10391|	dorsal thalamus|DTH|
anterior (rostral) cingulate (medial prefrontal) cortex|	MFC|	12157|	cingulate gyrus, rostral (anterior) part|	CgGr|
dorsolateral prefrontal cortex|	DFC|	12116|	middle frontal gyrus|	MFG|
orbital frontal cortex|	OFC|	12123|	anterior intermediate orbital gyrus|	AOrG|
lateral ganglionic eminence|	LGE|	NA|	NA|	NA|
inferolateral temporal cortex (area TEv, area 20)|	ITC|	12142|	inferior temporal gyrus|	ITG|
hippocampus (hippocampal formation)|	HIP|	12170|	hippocampal gyrus (formation)|	HiF|
ventrolateral prefrontal cortex|	VFC|	12117|	inferior frontal gyrus	IFG|
parietal neocortex|	PCx|	NA|	NA|	NA|
temporal neocortex|	TCx|	NA|	NA|	NA|
primary auditory cortex (core)|	A1C|	12144|	transverse| temporal gyrus (Heschl's gyrus)|	TTG|
primary visual cortex (striate cortex, area V1/17)|	V1C|	12150|	cuneus|	Cun|
striatum|	STR|	10333|	striatum|	STR|
primary motor cortex (area M1, area 4)|	M1C|	12114|	precentral gyrus|	PrCG|
posteroventral (inferior) parietal cortex	|IPC	|12135	|supramarginal gyrus	|SMG|
primary somatosensory cortex (area S1, areas 3,1,2)	|S1C|	12132	|postcentral gyrus|	PoCG|
cerebellum	|CB|	NA	|NA	|NA|
cerebellar cortex	|CBC	|10657	|cerebellar cortex	|CBC|
mediodorsal nucleus of thalamu|s	MD|	10398|	mediodorsal nucleus of thalamus|	MD|

Once this file has been created (using any tab/csv file editor), we save it as `mapping.csv`.


## Loading the data

```{r, eval = TRUE}
brain = readRDS("../../coldcuts_segmentations/ABA_Human_half/ABA_Human_half_seg.RDS")
bspan = read.csv("../../coldcuts_shiny/brainspan/expression_matrix.csv", header = FALSE, sep = ",")
coldata = read.csv("../../coldcuts_shiny/brainspan/columns_metadata.csv", header = TRUE, sep = ",")
rowdata = read.csv("../../coldcuts_shiny/brainspan/rows_metadata.csv", header = TRUE, sep = ",")
mapping = read.csv("../../coldcuts_shiny/brainspan/mapping.csv", header = TRUE, sep = ",")
```

The data needs some wrangling. We remove the structures with `NA` values, i.e. those that do not have a counterpart in our segmentation (in this case for developmental reasons)l; then we remove the corresponding samples.

Moreover, we have a duplicate structure, which will cause a conflict. The mediodorsal nucleus of the thalamus is present both as the structure itself, and as part of the dorsal thalamus, so we remove it from the thalamus and we keep it as its own structure.

```{r}
bspan_mapping = lapply(mapping$segmenation_acronym[!is.na(mapping$segmenation_acronym)], function(x) seg_select_str(brain, x))
names(bspan_mapping) = mapping$acronym[!is.na(mapping$segmenation_acronym)]

# Check duplicates:
if(any(duplicated(unlist(bspan_mapping)))) dup = unlist(bspan_mapping)[which(duplicated(unlist(bspan_mapping)))]

   lapply(bspan_mapping, function(x) x %in% dup)
   
# Remove the duplicate   
   bspan_mapping$DTH = bspan_mapping$DTH[bspan_mapping$DTH != "MD"] 
   
coldata$sample_id = paste0(coldata$donor_name, "_", gsub(coldata$age, pattern = " ", replacement = "_"), "_", coldata$structure_acronym)
rownames(coldata) = coldata$sample_id
colnames(bspan) = coldata$sample_id
rownames(bspan) = rowdata$ensembl_gene_id

coldata = coldata[coldata$structure_acronym %in% names(bspan_mapping),]
bspan = bspan[,coldata$sample_id]
```

## Creating the assay

Now we have everything to create the assay, add it to the segmentation object and add the projection for this assay:

```{r}
bspan_assay = new("segmentationAssay", values = as.matrix(bspan), mapping = bspan_mapping, sampledata = coldata)
 
brain = seg_assay_add(brain, assay = bspan_assay, name = "brainspan")

brain = seg_projection_add(segmentation = brain, name = "brainspan", structures = unlist(bspan_mapping), planes_chosen = "sagittal")

```

## Plotting the assay

We have two options. One is the simple feature plot, in which we choose a `feature` (gene), a grouping variable in the first element of `by` (age), and a level of the variable in the second element of `by` (11 yrs). Samples in the grouping variable will be aggregated using a function specified in `aggr_fun`, in this case `mean`.

```{r}
seg_feature_plot(brain, feature = "ENSG00000145335", by = c("age", "11 yrs"), assay = "brainspan", aggr_fun = "mean")
```


The other option is the complex plot, in which the feature is shown across all the different levels of the `by` argument:


```{r}
seg_feature_complex_plot(brain, feature = "ENSG00000145335", by = c("age", "11 yrs"), assay = "brainspan", aggr_fun = "mean")
```


Other aggregations are possible, e.g. by gender (F/M):

```{r}
seg_feature_complex_plot(brain, feature = "ENSG00000145335", by = c("gender", "F"), assay = "brainspan", aggr_fun = "mean")
```

